### PR TITLE
drug synergy score calculation added

### PR DESCRIPTION
**Pull request by runjin326**
# Drug synergy score calculation added

## Description

This module is written to implement the synergy from gene expression and network mining (SynGeNet) method described in [this paper](https://www.nature.com/articles/s41540-019-0085-4) with custom modifications.

Fixes # 1202
https://github.com/d3b-center/bixu-tracker/issues/1202

## Type of review for this PR
- [x] Scientific / output review
- [x] Code quality review

note: scientific/output review requires the lead scientists or analysts of the project to be more focus on the output of the PR where code quality review requires reviewer to make sure the PR is following the [contributing guideline](https://www.notion.so/d3b/contributing-721d172616b24ff395a64c18c58534f5) 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

I ran the script locally and all expected results were generated without errors.

**Task link/Screenshot/Terminal returns**:

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have emojis in my PR and commits 😉

### CODE REVIEW

File: code/drug_synergy/02-drug_synergy_score_calc.R
Comment (komalsrathi) [reply code/drug_synergy/02-drug_synergy_score_calc.R:46]: Agreed. 

Right now I have set up the code such that all patient level scripts go here: https://github.com/d3b-center/OMPARE/tree/master/code/patient_level_analyses and all corresponding functions go here: https://github.com/d3b-center/OMPARE/tree/master/code/patient_level_analyses/utils but they are mostly individual R scripts. Because you have been adding modules with multiple R scripts, we can keep it the way you have them structured but just moving them under `code/module_name/utils` (in this case `code/drug_synergy/utils`) and sourcing in the R scripts would be ideal.

Comment (runjin326) [reply code/drug_synergy/02-drug_synergy_score_calc.R]: @komalsrathi, thanks! I have now moved the scripts to utils ;)

Comment (komalsrathi) [reply code/drug_synergy/02-drug_synergy_score_calc.R]: @runjin326 so what you had was sort of correct, you just needed to move all the "functions" in the `code/module-name/utils/` folder and source them in your `code/module-name/01-script.R`, `code/module-name/02-script.R`.... Just like how you would for OT or OpenPBTA modules.

Comment (komalsrathi) [reply code/drug_synergy/02-drug_synergy_score_calc.R]: An example is [enrichment script](https://github.com/d3b-center/OMPARE/blob/master/code/patient_level_analyses/gsea_enrichment.R) which takes input parameters that can be tweaked, [sources the required function ](https://github.com/d3b-center/OMPARE/blob/master/code/patient_level_analyses/gsea_enrichment.R#L24) and [calls](https://github.com/d3b-center/OMPARE/blob/master/code/patient_level_analyses/gsea_enrichment.R#L130-L139) it.

Comment (runjin326) [reply code/drug_synergy/02-drug_synergy_score_calc.R]: Gotcha! Is this better? Basically I just moved the function and kept the scripts as they were.

File: code/drug_synergy/02-drug_synergy_score_calc.R
Comment (aadamk) [inline code/drug_synergy/02-drug_synergy_score_calc.R:71]: @runjin326 - is this weighting pulled from SynGeNet algorithm?

File: code/drug_synergy/utils/get_synergy_score.R
Comment (aadamk) [inline code/drug_synergy/utils/get_synergy_score.R:16]: @runjin326 - can you add a Roxygen skeleton here that defines the function's purpose and parameter definitions?

File: code/drug_synergy/02-drug_synergy_score_calc.R
Comment (runjin326) [reply code/drug_synergy/02-drug_synergy_score_calc.R:71]: Yes, this `(1 + (1.0 - (c(1:nDrug)/nDrug)))` calculation is from them but it is also slightly different. For them, they ranked the drugs based on how many targets a particular drug has in the subnetwork, take top 100 and give them weights between 2 and 1. For all other drugs that have >=2 targets in the subnetwork, they give them weight of 1. For us, [based on our discussion](https://github.com/d3b-center/bixu-tracker/issues/1166), we said we could use WTCS. However, to keep the scale the same to allow batch-to-batch comparison, I ranked by WTCS, and then give them weights between 2 to 1 based on the rank. Please let me know whether this is acceptable. I can modify if not.

File: code/drug_synergy/utils/get_synergy_score.R
Comment (runjin326) [reply code/drug_synergy/utils/get_synergy_score.R:34]: Added! Please let me know whether it's clear enough.

File: code/drug_synergy/02-drug_synergy_score_calc.R
Comment (aadamk) [inline code/drug_synergy/02-drug_synergy_score_calc.R:53]: with the subnetwork file you are reading in, would this background graph step run on every network module positively associated with the patient of interest (e.g. if there are multiple modules positively associated with the patient of interest)?

Comment (runjin326) [reply code/drug_synergy/02-drug_synergy_score_calc.R:53]: Currently I chose the network that has the highest positive score so this would only be one network. Do we want to do all networks?

Comment (aadamk) [reply code/drug_synergy/02-drug_synergy_score_calc.R:53]: Yes, expanding to all subnetworks would be the goal. The synergy output files you generate (e.g. `gtex_qSig_synergy_score.tsv`, etc) are nicely formatted, and adding the module number (e.g. `module` column with `Mx` designation) as an additional field would be informative. The final file could then continue to be ordered by synergy score in decreasing order as you have it.

Comment (runjin326) [reply code/drug_synergy/02-drug_synergy_score_calc.R:53]: Sure! Will do that :)

**Comment by runjin326** [comment]
@aadamk, I have now modified the code to loop through all modules that are positively correlated with the cluster and added a column in the output file indicating which module it is. Let me know whether it looks good - or if additional modification is required.

**Comment by aadamk** [comment]
> @aadamk, I have now modified the code to loop through all modules that are positively correlated with the cluster and added a column in the output file indicating which module it is. Let me know whether it looks good - or if additional modification is required.

Thank you @runjin326. Code changes look good. This will be a great addition to the reports. 
One element I missed is to retain the MOA category in the output files (the `*synergy_score.tsv` files) for each drug. So 1 column for each (MOA1 for drug1 and MOA2 for drug2 for each synergy pair). 
Once that is included, we can merge this in if @komalsrathi finds that it runs smoothly.

**Comment by runjin326** [comment]
@aadamk, I have now added the MOA and the target gene symbol for the drugs - the reason for doing that is for some drugs, depending on which gene target it has, it has different MOA:
For example, THIORIDAZINE  has MOA of `Serotonin 2a (5-HT2a) receptor antagonist` when its target is `HTR2A` and MOA of `D2-like dopamine receptor antagonist` when its target is `DRD3`. Hence I included target gene symbol as well to avoid confusion. 
There are drugs that have same MOA regardless its targets, those entries would just be duplicated with differences in target genes. 
Please let me know whether it makes sense and whether additional updates are needed :)

**Review by aadamk — APPROVED** [review]
Looks good, approving. Thank you.

**Review by komalsrathi — APPROVED** [review]
Looks good to me, approving.

**Comment by komalsrathi** [comment]
@runjin326 this looks good to me as is, I am going to make a few changes on my end i.e. replacing bash with an Rscript that calls the two scripts, removing hardcoded patient identifier and changing some paths i.e. cemitools directory for a patient lives in `OMPARE/results/PNOC008-XX/CEMITools` and all other reference files downloaded from s3 are under `OMPARE/data/reference` but thought it would be faster if I made those changes. Thanks for adding this!!

**Comment by runjin326** [comment]
@komalsrathi, gotcha! Thanks so much for making those modifications ;)

**Pull request merged**
